fw/http.c

```diff
@@ -1081,26 +1086,25 @@ void
 tfw_h2_resp_fwd(TfwHttpResp *resp)
 {
 	TfwHttpReq *req = resp->req;
-	TfwH2Ctx *ctx = tfw_h2_context(req->conn);
+	TfwConn *conn = req->conn;
+	bool resp_in_xmit = !!(SS_SKB_CB(((TfwMsg *)resp)->skb_head)->opaque_data);
 
-	tfw_connection_get(req->conn);
+	tfw_connection_get(conn);
 	do_access_log(resp);
 
-	if (tfw_cli_conn_send((TfwCliConn *)req->conn, (TfwMsg *)resp)) {
+	if (tfw_cli_conn_send((TfwCliConn *)conn, (TfwMsg *)resp)) {
 		T_DBG("%s: cannot send data to client via HTTP/2\n", __func__);
 		TFW_INC_STAT_BH(serv.msgs_otherr);
-		tfw_connection_close(req->conn, true);
-	}
-	else {
+		tfw_connection_close(conn, true);
+		/* We can't send response, so we can should free it here. */
+		resp_in_xmit = false;
+	} else {
 		TFW_INC_STAT_BH(serv.msgs_forwarded);
 		tfw_inc_global_hm_stats(resp->status);
 	}
 
-	tfw_connection_put(req->conn);
-
-	tfw_hpack_enc_release(&ctx->hpack, resp->flags);
-
-	tfw_http_resp_pair_free(req);
+	if (!resp_in_xmit)
+		tfw_http_resp_pair_free_and_put_conn(resp);
```

Comment:
It appears that just a simple two-line statement is fine, without the introduction of a special function that is only used once. 
```C
if (!resp_in_xmit) {
    tfw_connection_put(req->conn);
    tfw_http_resp_pair_free(req);
}
```

Reply:
Also, I am curious - why `resp` was chosen as an argument? That's in dissonance with the original function. Perhaps, a more meaningful change might be renaming the original function to `tfw_http_req_pair_free()`. :-) Frankly, I don't recall why is was named exactly that way. The meaning is still the same: _request and response pair_, _request and the corresponding response_.